model/sharing/sharing.go

```diff
@@ -1056,206 +1083,423 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
+			})
+			return checks
 		}
 
-		if !s.Active {
-			return nil
+		var ms *Sharing
+		if ms, err = FindSharing(member, s.ID()); err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":    s.SID,
+				"type":  "missing_sharing_for_member",
+				"error": err.Error(),
+			})
+			return checks
 		}
 
-		if s.Owner {
-			for i, m := range s.Members {
-				if i == 0 || m.Status != MemberStatusReady {
-					continue
-				}
-				if s.Credentials[i-1].Client == nil {
-					checks = append(checks, map[string]interface{}{
-						"id":     s.SID,
-						"type":   "missing_oauth_client",
-						"member": i,
-						"owner":  true,
-					})
+		if ms.ReadOnly() {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "read_only_member",
+				"member": m.Instance,
+			})
+			return checks
+		}
+
+		memberDocs, err := FindMatchingDocs(member, *rule)
+		if err != nil {
+			checks = append(checks, map[string]interface{}{
+				"id":     s.SID,
+				"type":   "missing_matching_docs_for_member",
+				"member": m.Instance,
+				"error":  err.Error(),
+			})
+			return checks
+		}
+
+		var isMemberDiskFull bool
+		if usage, err := member.VFS().DiskUsage(); err != nil {
+			isMemberDiskFull = usage >= member.VFS().DiskQuota()
+		}
+
+		// Build a map of member docs with their owner's counterpart ids
+		memberKey := s.Credentials[i].XorKey
+		memberDocsByID := make(map[string]couchdb.JSONDoc)
+		for _, doc := range memberDocs {
+			memberDocsByID[XorID(doc.ID(), memberKey)] = doc
+		}
+
+		for _, ownerDoc := range ownerDocs {
+			if memberDoc, found := memberDocsByID[ownerDoc.ID()]; found {
+				if ownerDoc.Rev() != memberDoc.Rev() {
+					if ownerDoc.ShortRev() > memberDoc.ShortRev() && isMemberDiskFull {
+						checks = append(checks, map[string]interface{}{
+							"id":       s.SID,
+							"type":     "disk_quota_exceeded",
+							"instance": m.Instance,
+						})
+						return checks
+					} else if ownerDoc.ShortRev() < memberDoc.ShortRev() && isOwnerDiskFull {
+						checks = append(checks, map[string]interface{}{
+							"id":       s.SID,
+							"type":     "disk_quota_exceeded",
+							"instance": inst.Domain,
+						})
+						return checks
+					} else {
+						checks = append(checks, map[string]interface{}{
+							"id":        s.SID,
+							"type":      "invalid_doc_rev",
+							"member":    m.Instance,
+							"ownerDoc":  ownerDoc,
+							"memberRev": memberDoc.Rev(),
+						})
+					}
+				} else {
+					// It's unnecessary to run these checks if both docs don't
+					// have the same revision in the first place.
+
+					if ownerDoc.M["name"] != memberDoc.M["name"] {
+						checks = append(checks, map[string]interface{}{
+							"id":         s.SID,
+							"type":       "invalid_doc_name",
+							"member":     m.Instance,
+							"ownerDoc":   ownerDoc,
+							"memberName": memberDoc.M["name"],
+						})
+					}
+
+					if ownerDoc.M["dir_id"] != memberDoc.M["dir_id"] {
```

Comment:
2 cases to manage:

1. the id has been XORed, so we need to redo a XOR to compare them
2. I'm not sure, but I think it will give errors for files/folders that are directly inside the shared directory (not in a sub-directory)

Reply:
Ah right, I forgot ids were XORed. And you're probably right about 2.